crates/store/src/db/sql.rs

```diff
@@ -604,6 +606,84 @@ pub fn select_notes_since_block_by_tag_and_sender(
     Ok(res)
 }
 
+/// Select notes matching the tag search criteria using the given [Connection].
+///
+/// # Returns
+///
+/// - Empty vector if no tag created after `block_num` match `tags`.
+/// - Otherwise, notes which the 16 high bits match `tags`, or the `sender` is one of the
+///   `account_ids`.
+///
+/// # Note
+///
+/// This method returns notes from a single block. To fetch all notes up to the chain tip,
+/// multiple requests are necessary.
+pub fn select_notes_since_block_by_tag(
+    conn: &mut Connection,
+    tags: &[u32],
+    block_num: BlockNumber,
+) -> Result<Vec<NoteRecord>> {
+    let tags: Vec<Value> = tags.iter().copied().map(u32_to_value).collect();
+
+    let mut stmt = conn.prepare(
+        "
+        SELECT
+            block_num,
+            batch_index,
+            note_index,
+            note_id,
+            note_type,
+            sender,
+            tag,
+            aux,
+            merkle_path,
+            details
+        FROM
+            notes
+        -- find the next block which contains at least one note with a matching tag
+        WHERE
+            tag IN rarray(?1) AND
+            block_num > ?2
+        ORDER BY
+            block_num ASC
+        LIMIT
+            1;
```

Comment:
The similar query in `select_notes_since_block_by_tag_and_sender` above contains a level of nesting -- I'm not sure why the nested select is there?

Reply:
Without nested select, wouldn't we just get a single row returned from here (the `LIMIT 1` would apply to the whole query, right?)?

What we want to do is get all relevant notes from the first block which hash at least one relevant note. So, we first need to find this block (this is done via the inner select) and then get all matching notes in this blocks (that's the outer select).

Reply:
Ah. Yes, I had (incorrectly) assumed that the array contained all the notes. But of course the array is just all of the tags for a single note. Thanks!